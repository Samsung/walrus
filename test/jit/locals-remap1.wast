(module
(func (result i32) (local i32 i32)
  i32.const 1
  local.set 0

  i32.const 5
  if
    i32.const 1
    local.set 0

    local.get 0
    drop
  else
    i32.const 1
    local.set 1
  end

  local.get 1
  drop

  local.get 0
)

(func (local i32)
  i32.const 1
  drop

  i32.const 1
  local.set 0

  loop
    loop
      i32.const 1
      local.set 0

      i32.const 0
      br_if 1

      i32.const 0
      br_if 0
    end
  end
)

(func (local i32)
  loop
    i32.const 0
    br_if 0
  end

  loop
    loop
      i32.const 0
      br_if 1

      i32.const 0
      if
        i32.const 1
        local.set 0
      else
        local.get 0
        br_if 1
      end
    end
  end

  loop
    i32.const 0
    br_if 0
  end
)

(func (local i32)
  loop
    i32.const 0
    if
      i32.const 0
      local.set 0
    else
      i32.const 0
      local.set 0
    end
    local.get 0
    br_if 0
  end
)

(func (local i32 i32 i32)
  i32.const 0
  local.set 2
  local.get 2
  local.set 1
  local.get 1
  local.set 0
  local.get 0
  drop
)

(func (param i32 i32) (local i32 i32 i32)
  i32.const 0
  local.set 4
  local.get 4
  local.set 0
)
)
