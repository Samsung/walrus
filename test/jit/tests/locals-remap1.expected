[[[[[[[  Function   0  ]]]]]]]
Local 1: mapped to: 0 range [BEGIN - 13]
Local 0: mapped to: 1 range [1 - 15]
Number of locals reduced from 2 to 2
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 1
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: br_unless
  Jump to: 9
  Param[0]: ref:2(0)
4: Opcode: i32.const
  Result[0]: i32
5: Opcode: local.set
  Index: 1
  Param[0]: ref:4(0)
6: Opcode: local.get
  Index: 1
  Result[0]: i32
7: Opcode: drop
  Param[0]: ref:6(0)
8: Opcode: br
  Jump to: 12
9: Label: result_count: 0 preserved_count: 0
  Jump from: 3
10: Opcode: i32.const
  Result[0]: i32
11: Opcode: local.set
  Index: 0
  Param[0]: ref:10(0)
12: Label: result_count: 0 preserved_count: 0
  Jump from: 8
13: Opcode: local.get
  Index: 0
  Result[0]: i32
14: Opcode: drop
  Param[0]: ref:13(0)
15: Opcode: local.get
  Index: 1
  Result[0]: i32
16: Opcode: return
  Param[0]: ref:15(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: br_unless
  Jump to: 9
  Param[0]: i32 const:2
4: Opcode: i32.const
  Result[0]: i32 unused
5: Opcode: local.set
  Addr: sp[8]
  Param[0]: i32 const:4
6: Opcode: local.get
  Addr: sp[8]
  Result[0]: i32 unused
7: Opcode: drop
  Param[0]: i32 sp[8]
8: Opcode: br
  Jump to: 12
9: Label: result_count: 0 preserved_count: 0
  Jump from: 3
10: Opcode: i32.const
  Result[0]: i32 unused
11: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:10
12: Label: result_count: 0 preserved_count: 0
  Jump from: 8
13: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
14: Opcode: drop
  Param[0]: i32 sp[4]
15: Opcode: local.get
  Addr: sp[8]
  Result[0]: i32 sp[0]
16: Opcode: return
  Param[0]: i32 sp[0]

[[[[[[[  Function   1  ]]]]]]]
Local 0: mapped to: 0 range [3 - 10]
Number of locals reduced from 1 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: drop
  Param[0]: ref:0(0)
2: Opcode: i32.const
  Result[0]: i32
3: Opcode: local.set
  Index: 0
  Param[0]: ref:2(0)
4: Label: result_count: 0 preserved_count: 0
  Jump from: 8
  Jump from: 10
5: Opcode: i32.const
  Result[0]: i32
6: Opcode: local.set
  Index: 0
  Param[0]: ref:5(0)
7: Opcode: i32.const
  Result[0]: i32
8: Opcode: br_if
  Jump to: 4
  Param[0]: ref:7(0)
9: Opcode: i32.const
  Result[0]: i32
10: Opcode: br_if
  Jump to: 4
  Param[0]: ref:9(0)
11: Opcode: return
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: drop
  Param[0]: i32 const:0
2: Opcode: i32.const
  Result[0]: i32 unused
3: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:2
4: Label: result_count: 0 preserved_count: 0
  Jump from: 8
  Jump from: 10
5: Opcode: i32.const
  Result[0]: i32 unused
6: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:5
7: Opcode: i32.const
  Result[0]: i32 unused
8: Opcode: br_if
  Jump to: 4
  Param[0]: i32 const:7
9: Opcode: i32.const
  Result[0]: i32 unused
10: Opcode: br_if
  Jump to: 4
  Param[0]: i32 const:9
11: Opcode: return

[[[[[[[  Function   2  ]]]]]]]
Local 0: mapped to: 0 range [BEGIN - 13]
Number of locals reduced from 1 to 1
0: Label: result_count: 0 preserved_count: 0
  Jump from: 2
1: Opcode: i32.const
  Result[0]: i32
2: Opcode: br_if
  Jump to: 0
  Param[0]: ref:1(0)
3: Label: result_count: 0 preserved_count: 0
  Jump from: 5
  Jump from: 13
4: Opcode: i32.const
  Result[0]: i32
5: Opcode: br_if
  Jump to: 3
  Param[0]: ref:4(0)
6: Opcode: i32.const
  Result[0]: i32
7: Opcode: br_unless
  Jump to: 11
  Param[0]: ref:6(0)
8: Opcode: i32.const
  Result[0]: i32
9: Opcode: local.set
  Index: 0
  Param[0]: ref:8(0)
10: Opcode: br
  Jump to: 14
11: Label: result_count: 0 preserved_count: 0
  Jump from: 7
12: Opcode: local.get
  Index: 0
  Result[0]: i32
13: Opcode: br_if
  Jump to: 3
  Param[0]: ref:12(0)
14: Label: result_count: 0 preserved_count: 0
  Jump from: 10
  Jump from: 16
15: Opcode: i32.const
  Result[0]: i32
16: Opcode: br_if
  Jump to: 14
  Param[0]: ref:15(0)
17: Opcode: return
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Label: result_count: 0 preserved_count: 0
  Jump from: 2
1: Opcode: i32.const
  Result[0]: i32 unused
2: Opcode: br_if
  Jump to: 0
  Param[0]: i32 const:1
3: Label: result_count: 0 preserved_count: 0
  Jump from: 5
  Jump from: 13
4: Opcode: i32.const
  Result[0]: i32 unused
5: Opcode: br_if
  Jump to: 3
  Param[0]: i32 const:4
6: Opcode: i32.const
  Result[0]: i32 unused
7: Opcode: br_unless
  Jump to: 11
  Param[0]: i32 const:6
8: Opcode: i32.const
  Result[0]: i32 unused
9: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:8
10: Opcode: br
  Jump to: 14
11: Label: result_count: 0 preserved_count: 0
  Jump from: 7
12: Opcode: local.get
  Addr: sp[0]
  Result[0]: i32 unused
13: Opcode: br_if
  Jump to: 3
  Param[0]: i32 sp[0]
14: Label: result_count: 0 preserved_count: 0
  Jump from: 10
  Jump from: 16
15: Opcode: i32.const
  Result[0]: i32 unused
16: Opcode: br_if
  Jump to: 14
  Param[0]: i32 const:15
17: Opcode: return

[[[[[[[  Function   3  ]]]]]]]
Local 0: mapped to: 0 range [4 - 10]
Number of locals reduced from 1 to 1
0: Label: result_count: 0 preserved_count: 0
  Jump from: 11
1: Opcode: i32.const
  Result[0]: i32
2: Opcode: br_unless
  Jump to: 6
  Param[0]: ref:1(0)
3: Opcode: i32.const
  Result[0]: i32
4: Opcode: local.set
  Index: 0
  Param[0]: ref:3(0)
5: Opcode: br
  Jump to: 9
6: Label: result_count: 0 preserved_count: 0
  Jump from: 2
7: Opcode: i32.const
  Result[0]: i32
8: Opcode: local.set
  Index: 0
  Param[0]: ref:7(0)
9: Label: result_count: 0 preserved_count: 0
  Jump from: 5
10: Opcode: local.get
  Index: 0
  Result[0]: i32
11: Opcode: br_if
  Jump to: 0
  Param[0]: ref:10(0)
12: Opcode: return
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Label: result_count: 0 preserved_count: 0
  Jump from: 11
1: Opcode: i32.const
  Result[0]: i32 unused
2: Opcode: br_unless
  Jump to: 6
  Param[0]: i32 const:1
3: Opcode: i32.const
  Result[0]: i32 unused
4: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:3
5: Opcode: br
  Jump to: 9
6: Label: result_count: 0 preserved_count: 0
  Jump from: 2
7: Opcode: i32.const
  Result[0]: i32 unused
8: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:7
9: Label: result_count: 0 preserved_count: 0
  Jump from: 5
10: Opcode: local.get
  Addr: sp[0]
  Result[0]: i32 unused
11: Opcode: br_if
  Jump to: 0
  Param[0]: i32 sp[0]
12: Opcode: return

[[[[[[[  Function   4  ]]]]]]]
Local 2: mapped to: 0 range [1 - 2]
Local 1: mapped to: 0 range [3 - 4]
Local 0: mapped to: 0 range [5 - 6]
Number of locals reduced from 3 to 1
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: local.get
  Index: 0
  Result[0]: i32
3: Opcode: local.set
  Index: 0
  Param[0]: ref:2(0)
4: Opcode: local.get
  Index: 0
  Result[0]: i32
5: Opcode: local.set
  Index: 0
  Param[0]: ref:4(0)
6: Opcode: local.get
  Index: 0
  Result[0]: i32
7: Opcode: drop
  Param[0]: ref:6(0)
8: Opcode: return
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 const:0
2: Opcode: local.get
  Addr: sp[0]
  Result[0]: i32 unused
3: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 sp[0]
4: Opcode: local.get
  Addr: sp[0]
  Result[0]: i32 unused
5: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 sp[0]
6: Opcode: local.get
  Addr: sp[0]
  Result[0]: i32 unused
7: Opcode: drop
  Param[0]: i32 sp[0]
8: Opcode: return

[[[[[[[  Function   5  ]]]]]]]
Local 0: mapped to: 0 range [BEGIN - 3]
Local 1: mapped to: 1 unused
Local 2: mapped to: 0 unused
Local 3: mapped to: 0 unused
Local 4: mapped to: 1 range [1 - 2]
Number of locals reduced from 5 to 2
0: Opcode: i32.const
  Result[0]: i32
1: Opcode: local.set
  Index: 1
  Param[0]: ref:0(0)
2: Opcode: local.get
  Index: 1
  Result[0]: i32
3: Opcode: local.set
  Index: 0
  Param[0]: ref:2(0)
4: Opcode: return
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i32.const
  Result[0]: i32 unused
1: Opcode: local.set
  Addr: sp[4]
  Param[0]: i32 const:0
2: Opcode: local.get
  Addr: sp[4]
  Result[0]: i32 unused
3: Opcode: local.set
  Addr: sp[0]
  Param[0]: i32 sp[4]
4: Opcode: return

